File: itcloud/src/main/java/com/itinfo/itcloud/controller/network/NetworkController.java
package com.itinfo.itcloud.controller.network;

import com.itinfo.itcloud.model.computing.PermissionVo;
import com.itinfo.itcloud.model.create.NetworkCreateVo;
import com.itinfo.itcloud.model.error.CommonVo;
import com.itinfo.itcloud.model.network.*;
import com.itinfo.itcloud.service.network.ItNetworkService;
import io.swagger.annotations.Api;
import io.swagger.annotations.ApiImplicitParam;
import io.swagger.annotations.ApiOperation;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Controller
@Slf4j
@RequiredArgsConstructor
@Api(tags = "Network")
@RequestMapping("/networks")
public class NetworkController {
	private final ItNetworkService networkService;


	@GetMapping
	@ApiOperation(value = "네트워크 목록", notes = "전체 네트워크 목록을 조회한다")
	@ResponseBody
	public List<NetworkVo> networks(){
		log.info("--- Network 목록");
		return networkService.getList();
	}

	@GetMapping("/settings")
	@ApiOperation(value = "네트워크 생성 창", notes = "네트워크 생성시 필요한 내용을 조회한다")
	@ResponseBody
	public List<NetworkDcClusterVo> setDcCluster(){
		log.info("--- Network 생성 창");
		return networkService.setDcCluster();
	}

	@PostMapping
	@ApiOperation(value = "네트워크 생성", notes = "네트워크를 생성한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public CommonVo<Boolean> addNetwork(@RequestBody NetworkCreateVo nVo){
		log.info("--- Network 생성");
        return networkService.addNetwork(nVo);
	}
	
	@GetMapping("/{id}/edit")
	@ApiOperation(value = "네트워크 수정창", notes = "선택된 네트워크의 정보를 조회한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public NetworkCreateVo setEditNetwork(@PathVariable String id){
		log.info("--- Network 편집 창");
		return networkService.setEditNetwork(id);
	}

	@PutMapping("/{id}")
	@ApiOperation(value = "네트워크 수정", notes = "네트워크를 수정한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public CommonVo<Boolean> editNetwork(@PathVariable String id,
										 @RequestBody NetworkCreateVo nVo){
		log.info("--- Network 편집");
        return networkService.editNetwork(nVo);
	}

	@DeleteMapping("/{id}")
	@ApiOperation(value = "네트워크 삭제", notes = "네트워크를 삭제한다")
	@ApiImplicitParam(name = "id", value = "네트워크 아이디", dataTypeClass = String.class)
	@ResponseBody
	public CommonVo<Boolean> deleteNetwork(@PathVariable String id){
		log.info("--- Network 삭제");
        return networkService.deleteNetwork(id);
	}



	// TODO:HELP
	// 네트워크 가져오기 창
	@GetMapping("/import/settings")
	@ResponseBody
	public NetworkImportVo setImportNetwork(){
		log.info("--- Network 가져오기 창");
		return networkService.setImportNetwork();
	}

	// 네트워크 가져오기
	@PostMapping("/import")
	public CommonVo<Boolean> importNw(){
		log.info("--- Network 가져오기");
		return networkService.importNetwork();
	}



	@GetMapping("/{id}")
	@ApiOperation(value = "네트워크 상세정보", notes = "네트워크의 상세정보를 조회한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public NetworkVo network(@PathVariable String id){
		log.info("--- Network 일반");
		return networkService.getNetwork(id);
	}


	@GetMapping("/{id}/vnics")
	@ApiOperation(value = "네트워크 vnic 프로파일 목록", notes = "선택된 네트워크의 vnic 프로파일 목록을 조회한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public List<VnicProfileVo> vnic(@PathVariable String id){
		log.info("--- Network vnic 프로파일");
		return networkService.getVnic(id);
	}

	// vnic 생성 창
	// TODO:HELP
	@GetMapping("/{id}/vnics/settings")
	@ResponseBody
	public VnicCreateVo setVnic(@PathVariable String id){
		log.info("--- Network vnic 프로파일 생성창");
		return networkService.setVnic(id);
	}

	// vnic 생성
	@PostMapping("/{id}/vnic")
	@ResponseBody
	public CommonVo<Boolean> addVnic(@PathVariable String id,
									 @RequestBody VnicCreateVo vcVo){
		log.info("--- Network vnic 프로파일 생성");
		return networkService.addVnic(id, vcVo);
	}

	// vnic 편집 창
	@GetMapping("/{id}/vnics/{vcId}/settings")
	@ResponseBody
	public VnicCreateVo setEditVnic(@PathVariable String id,
								 @PathVariable String vcId){
		log.info("--- Network vnic프로파일 편집창");
		return networkService.setEditVnic(id, vcId);
	}

	// vnic 편집
	@PutMapping("/{id}/vnics/{vcId}")
	@ResponseBody
	public CommonVo<Boolean> editVnic(@PathVariable String id,
									  @PathVariable String vcId,
									  @RequestBody VnicCreateVo vcVo){
		log.info("--- Network vnic프로파일 편집");
		return networkService.editVnic(id, vcId, vcVo);
	}

	// vnic 삭제
	@DeleteMapping("/{id}/vnics/{vcId}")
	@ResponseBody
	public CommonVo<Boolean> deleteVnic(@PathVariable String id,
										@PathVariable String vcId){
		log.info("--- Network vnic프로파일 삭제");
		return networkService.deleteVnic(id, vcId);
	}



	// 클러스터 목록
	@GetMapping("/{id}/clusters")
	@ApiOperation(value = "네트워크 클러스터 목록", notes = "선택된 네트워크의 클러스터 목록을 조회한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public List<NetworkClusterVo> cluster(@PathVariable String id){
		log.info("--- Network 클러스터");
		return networkService.getCluster(id);
	}
	
	// 클러스터 네트워크 관리창
	@GetMapping("/{id}/clusters/{cId}/settings")
	@ResponseBody
	public NetworkUsageVo getUsage(@PathVariable String id, 
								   @PathVariable String cId){
		log.info("--- Network 클러스터 네트워크 관리 창");
		return networkService.getUsage(id, cId);
	}
	
	
	

	@GetMapping("/{id}/hosts")
	@ApiOperation(value = "네트워크 호스트 목록", notes = "선택된 네트워크의 호스트 목록을 조회한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public List<NetworkHostVo> host(@PathVariable String id){
		log.info("--- Network 호스트");
		return networkService.getHost(id);
	}

	@GetMapping("/{id}/vms")
	@ApiOperation(value = "네트워크 가상머신 목록", notes = "선택된 네트워크의 가상머신 목록을 조회한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public List<NetworkVmVo> vm(@PathVariable String id){
		log.info("--- Network 가상머신");
		return networkService.getVm(id);
	}

	// 가상머신 nic 제거
	@DeleteMapping("/{id}/vms/{vmId}/{nicId}")
	@ResponseBody
	public CommonVo<Boolean> deleteVmNic(@PathVariable String id,
										 @PathVariable String vmId,
										 @PathVariable String nicId){
		log.info("--- Network 가상머신 nic 제거");
		return networkService.deleteVmNic(id, vmId, nicId);
	}

	@GetMapping("/{id}/templates")
	@ApiOperation(value = "네트워크 템플릿 목록", notes = "선택된 네트워크의 템플릿 목록을 조회한다")
	@ApiImplicitParam(name = "nVo", value = "네트워크", dataTypeClass = String.class)
	@ResponseBody
	public List<NetworkTemplateVo> template(@PathVariable String id){
		log.info("--- Network 템플릿");
		return networkService.getTemplate(id);
	}

	// 템플릿 제거
	@DeleteMapping("/{id}/templates/{tempId}/{nicId}")
	@ResponseBody
	public CommonVo<Boolean> deleteTempNic(@PathVariable String id,
										   @PathVariable String tempId,
										   @PathVariable String nicId){
		log.info("--- Network 템플릿 nic 제거");
		return networkService.deleteTempNic(id, tempId, nicId);
	}


	// 권한 목록
	@GetMapping("/{id}/permissions")
	@ApiOperation(value = "네트워크 권한 목록", notes = "선택된 네트워크의 권한 목록을 조회한다")
	@ApiImplicitParam(name = "id", value = "네트워크 아이디", dataTypeClass = String.class)
	@ResponseBody
	public List<PermissionVo> permission(@PathVariable String id){
		log.info("--- Network 권한");
		return networkService.getPermission(id);
	}


}

File: itcloud/src/main/java/com/itinfo/itcloud/controller/network/NetworkController.kt
package com.itinfo.itcloud.controller.network

import com.itinfo.common.LoggerDelegate
import com.itinfo.itcloud.controller.BaseController
import com.itinfo.itcloud.error.toException
import com.itinfo.itcloud.model.IdentifiedVo
import com.itinfo.util.ovirt.error.ErrorPattern
import com.itinfo.itcloud.model.computing.ClusterVo
import com.itinfo.itcloud.model.computing.DataCenterVo
import com.itinfo.itcloud.model.computing.HostVo
import com.itinfo.itcloud.model.computing.VmVo
import com.itinfo.itcloud.model.setting.PermissionVo
import com.itinfo.itcloud.model.network.NetworkVo
import com.itinfo.itcloud.model.network.VnicProfileVo
import com.itinfo.itcloud.model.network.NetworkTemplateVo
import com.itinfo.itcloud.model.network.OpenStackNetworkVo
import com.itinfo.itcloud.service.network.ItNetworkService
import com.itinfo.itcloud.service.network.ItVnicProfileService
import io.swagger.annotations.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*

@Controller
@Api(tags = ["Network"])
@RequestMapping("/api/v1/networks")
class NetworkController: BaseController() {
	@Autowired private lateinit var iNetwork: ItNetworkService

	@ApiOperation(
		httpMethod="GET",
		value="네트워크 목록 조회",
		notes="전체 네트워크 목록을 조회한다"
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun networks(): ResponseEntity<List<NetworkVo>> {
		log.info("/networks ... 네트워크 목록")
		return ResponseEntity.ok(iNetwork.findAll())
	}

	@ApiOperation(
		httpMethod="GET",
		value="네트워크 상세정보",
		notes="네트워크의 상세정보를 조회한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/{networkId}")
	@ResponseBody
	fun network(
		@PathVariable networkId: String? = null
	): ResponseEntity<NetworkVo?> {
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		log.info("/networks/{} ... 네트워크 상세정보", networkId)
		return ResponseEntity.ok(iNetwork.findOne(networkId))
	}

	@ApiOperation(
		httpMethod="POST",
		value="네트워크 생성",
		notes="네트워크를 생성한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="network", value="네트워크", dataTypeClass=NetworkVo::class, required=true, paramType="body")
	)
	@ApiResponses(
		ApiResponse(code = 201, message = "CREATED"),
		ApiResponse(code = 404, message = "NOT_FOUND")
	)
	@PostMapping
	@ResponseBody
	@ResponseStatus(HttpStatus.CREATED)
	fun add(
		@RequestBody network: NetworkVo? = null
	): ResponseEntity<NetworkVo?> {
		if (network == null)
			throw ErrorPattern.NETWORK_VO_INVALID.toException()
		log.info("/networks ... 네트워크 생성")
		return ResponseEntity.ok(iNetwork.add(network))
	}

	@ApiOperation(
		httpMethod="PUT",
		value="네트워크 편집",
		notes="네트워크를 편집한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
		ApiImplicitParam(name="network", value="네트워크", dataTypeClass=NetworkVo::class, paramType="body")
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@PutMapping("/{networkId}")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun update(
		@PathVariable networkId: String? = null,
		@RequestBody network: NetworkVo? = null,
	): ResponseEntity<NetworkVo?> {
		log.info("/networks/{} ... 네트워크 편집", networkId)
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		if (network == null)
			throw ErrorPattern.NETWORK_VO_INVALID.toException()
		return ResponseEntity.ok(iNetwork.update(network))
	}

	@ApiOperation(
		httpMethod="DELETE",
		value="네트워크 삭제",
		notes="네트워크를 삭제한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@DeleteMapping("/{networkId}")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun remove(
		@PathVariable networkId: String? = null,
	): ResponseEntity<Boolean> {
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		log.info("/networks/{} ... 네트워크 삭제", networkId)
		return ResponseEntity.ok(iNetwork.remove(networkId))
	}

	@ApiOperation(
		httpMethod="GET",
		value="네트워크 가져오기 창",
		notes="네트워크 가져오기 창 - 네트워크 공급자 목록을 가져온다"
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/import/settings")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun importProviders(): ResponseEntity<IdentifiedVo?> {
		log.info("/networks/import/settings ... 네트워크 가져오기 창")
		return ResponseEntity.ok(iNetwork.findNetworkProviderFromNetwork())
	}

	@ApiOperation(
		httpMethod="GET",
		value="네트워크 가져오기 창",
		notes="네트워크 가져오기 창 - 네트워크 공급자가 가지고 있는 네트워크 목록을 가져온다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="providerId", value="공급자 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/import/settings/{providerId}")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun providers(
		@PathVariable providerId: String? = null
	): ResponseEntity<List<OpenStackNetworkVo?>> {
		if (providerId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_NOT_FOUND.toException()
		log.info("/networks/import/setting/{} ... 네트워크 가져오기 창", providerId)
		return ResponseEntity.ok(iNetwork.findAllOpenStackNetworkFromNetworkProvider(providerId))
	}

	@ApiOperation(
		httpMethod="GET",
		value="네트워크 가져오기 창",
		notes="네트워크 가져오기 창 - 네트워크를 가져올 수 있는 데이터센터 목록"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="openstackNetworkId", value="공급자 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/import/datacenters/{openstackNetworkId}")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun providersDatacenter(
		@PathVariable openstackNetworkId: String? = null
	): ResponseEntity<List<DataCenterVo?>> {
		if (openstackNetworkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_NOT_FOUND.toException()
		log.info("/networks/import/settingDatacenter/{} ... 네트워크 가져오기 창", openstackNetworkId)
		return ResponseEntity.ok(iNetwork.findAllDataCentersFromNetwork(openstackNetworkId))
	}

	@ApiOperation(
		httpMethod="POST",
		value="네트워크 가져오기",
		notes="네트워크를 가져온다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="openstackNetwork", value="네트워크", dataTypeClass=Array<OpenStackNetworkVo>::class, required=true, paramType="body")
	)
	@ApiResponses(
		ApiResponse(code = 201, message = "CREATED"),
		ApiResponse(code = 404, message = "NOT_FOUND")
	)
	@PostMapping("/import")
	@ResponseBody
	@ResponseStatus(HttpStatus.CREATED)
	fun import(
		@RequestBody openStackNetworkVos: List<OpenStackNetworkVo>? = null
	): ResponseEntity<Boolean?> {
		if (openStackNetworkVos == null)
			throw ErrorPattern.NETWORK_VO_INVALID.toException()
		log.info("/networks/import ... 네트워크 가져오기")
		return ResponseEntity.ok(iNetwork.importNetwork(openStackNetworkVos))
	}


	@ApiOperation(
		httpMethod="GET",
		value="네트워크 클러스터 목록",
		notes="선택된 네트워크의 클러스터 목록을 조회한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/{networkId}/clusters")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun clusters(
		@PathVariable networkId: String? = null,
	): ResponseEntity<List<ClusterVo>> {
		log.info("/networks/{}/clusters ... 네트워크 클러스터 목록", networkId)
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		return ResponseEntity.ok(iNetwork.findAllClustersFromNetwork(networkId))
	}

	@ApiOperation(
		httpMethod="GET",
		value="네트워크 호스트 목록",
		notes="선택된 네트워크의 호스트 목록을 조회한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/{networkId}/hosts")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun hosts(
		@PathVariable networkId: String? = null,
	): ResponseEntity<List<HostVo>> {
		log.info("/networks/{}/hosts ... 네트워크 호스트 목록", networkId)
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		return ResponseEntity.ok(iNetwork.findAllHostsFromNetwork(networkId))
	}

	@ApiOperation(
		httpMethod="GET",
		value = "네트워크 가상머신 목록",
		notes = "선택된 네트워크의 가상머신 목록을 조회한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/{networkId}/vms")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun vms(
		@PathVariable networkId: String? = null,
	): ResponseEntity<List<VmVo>> {
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		log.info("/networks/{}/vms ... 네트워크 가상머신 목록", networkId)
		return ResponseEntity.ok(iNetwork.findAllVmsFromNetwork(networkId))
	}

	@ApiOperation(
		httpMethod="GET",
		value = "네트워크 템플릿 목록",
		notes = "선택된 네트워크의 템플릿 목록을 조회한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/{networkId}/templates")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun templates(
		@PathVariable networkId: String? = null,
	): ResponseEntity<List<NetworkTemplateVo>> {
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		log.info("/networks/{}/templates ... 네트워크 템플릿 목록", networkId)
		return ResponseEntity.ok(iNetwork.findAllTemplatesFromNetwork(networkId))
	}

	@ApiOperation(
		httpMethod="GET",
		value = "네트워크 권한 목록",
		notes = "선택된 네트워크의 권한 목록을 조회한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/{networkId}/permissions")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	@Deprecated("사용안함")
	fun permissions(
		@PathVariable networkId: String? = null,
	): ResponseEntity<List<PermissionVo>> {
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		log.info("/networks/{}/permissions ... 네트워크 권한 목록", networkId)
		return ResponseEntity.ok(iNetwork.findAllPermissionsFromNetwork(networkId))
	}


	// region: vnicProfile
	@Autowired private lateinit var iVnic: ItVnicProfileService

	
	@ApiOperation(
		httpMethod="GET",
		value="전체 vnicProfile 목록 조회",
		notes="전체 vnicProfile 목록을 조회한다"
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/vnicProfiles")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun vnicProfiles(): ResponseEntity<List<VnicProfileVo>> {
		log.info("/vnicProfiles ... 네트워크 목록")
		return ResponseEntity.ok(iVnic.findAll())
	}

	@ApiOperation(
		httpMethod="GET",
		value="네트워크 vnic 프로파일 목록",
		notes="선택된 네트워크의 vnic 프로파일 목록을 조회한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/{networkId}/vnicProfiles")
	@ResponseBody
	fun vnics(
		@PathVariable networkId: String? = null,
	): ResponseEntity<List<VnicProfileVo>> {
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		log.info("/networks/{}/vnicProfiles ... 네트워크 vnic profile 목록", networkId)
		return ResponseEntity.ok(iVnic.findAllFromNetwork(networkId))
	}

	@ApiOperation(
		httpMethod="GET",
		value="네트워크 vnic 프로파일",
		notes="선택된 네트워크의 vnic 프로파일의 상세정보를 조회한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
		ApiImplicitParam(name="vnicProfileId", value="Vnic Profile ID", dataTypeClass=String::class, required=true, paramType="path"),
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@GetMapping("/{networkId}/vnicProfiles/{vnicProfileId}")
	@ResponseBody
	fun vnic(
		@PathVariable networkId: String? = null,
		@PathVariable vnicProfileId: String? = null,
	): ResponseEntity<VnicProfileVo?> {
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		if (vnicProfileId.isNullOrEmpty())
			throw ErrorPattern.VNIC_PROFILE_ID_NOT_FOUND.toException()
		log.info("/networks/{}/vnicProfiles/{} ... 네트워크 vnic profile", networkId, vnicProfileId)
		return ResponseEntity.ok(iVnic.findOne(vnicProfileId))
	}

	@ApiOperation(
		httpMethod="POST",
		value="vnic 프로파일 생성",
		notes="vnic 프로파일을 생성한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
		ApiImplicitParam(name="vnicProfile", value="Vnic Profile", dataTypeClass=VnicProfileVo::class, required=true, paramType="body")
	)
	@ApiResponses(
		ApiResponse(code = 201, message = "CREATED"),
		ApiResponse(code = 404, message = "NOT_FOUND")
	)
	@PostMapping("/{networkId}/vnicProfiles")
	@ResponseBody
	@ResponseStatus(HttpStatus.CREATED)
	fun addVnic(
		@PathVariable networkId: String? = null,
		@RequestBody vnicProfile: VnicProfileVo? = null
	): ResponseEntity<VnicProfileVo?> {
		if (vnicProfile == null)
			throw ErrorPattern.VNIC_PROFILE_VO_INVALID.toException()
		log.info("/networks/{}/vnicProfiles ... vnicProfile 생성", networkId)
		return ResponseEntity.ok(iVnic.add(vnicProfile))
	}

	@ApiOperation(
		httpMethod="PUT",
		value="vnic 프로파일 편집",
		notes="vnic 프로파일을 편집한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
		ApiImplicitParam(name="vnicProfileId", value="Vnic Profile ID", dataTypeClass=String::class, required=true, paramType="path"),
		ApiImplicitParam(name="vnicProfile", value="Vnic Profile", dataTypeClass=VnicProfileVo::class, paramType="body")
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@PutMapping("/{networkId}/vnicProfiles/{vnicProfileId}")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun updateVnic(
		@PathVariable networkId: String? = null,
		@PathVariable vnicProfileId: String? = null,
		@RequestBody vnicProfile: VnicProfileVo? = null,
	): ResponseEntity<VnicProfileVo?> {
		log.info("/networks/{}/vnicProfiles/{} ... vnicProfile 생성", networkId, vnicProfileId)
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		if (vnicProfileId.isNullOrEmpty())
			throw ErrorPattern.VNIC_PROFILE_ID_NOT_FOUND.toException()
		if (vnicProfile == null)
			throw ErrorPattern.VNIC_PROFILE_NOT_FOUND.toException()
		return ResponseEntity.ok(iVnic.update(vnicProfile))
	}

	@ApiOperation(
		httpMethod="DELETE",
		value="vnic 프로파일 삭제",
		notes="vnic 프로파일을 삭제한다"
	)
	@ApiImplicitParams(
		ApiImplicitParam(name="networkId", value="네트워크 ID", dataTypeClass=String::class, required=true, paramType="path"),
		ApiImplicitParam(name="vnicProfileId", value="Vnic Profile ID", dataTypeClass=String::class, required=true, paramType="path")
	)
	@ApiResponses(
		ApiResponse(code = 200, message = "OK")
	)
	@DeleteMapping("/{networkId}/vnicProfiles/{vnicProfileId}")
	@ResponseBody
	@ResponseStatus(HttpStatus.OK)
	fun removeVnic(
		@PathVariable networkId: String? = null,
		@PathVariable vnicProfileId: String? = null,
	): ResponseEntity<Boolean> {
		if (networkId.isNullOrEmpty())
			throw ErrorPattern.NETWORK_ID_NOT_FOUND.toException()
		if (vnicProfileId.isNullOrEmpty())
			throw ErrorPattern.VNIC_PROFILE_ID_NOT_FOUND.toException()
		log.info("/{}/vnicProfiles/{} ... vnic 프로파일 삭제", networkId, vnicProfileId)
		return ResponseEntity.ok(iVnic.remove(vnicProfileId))
	}


	// endregion

	companion object {
		private val log by LoggerDelegate()
	}

}